transformations/transformations/data_offload.py

```diff
@@ -176,3 +181,335 @@ def remove_openmp_pragmas(self, routine, targets):
+    key : str, optional
+        Overwrite the key that is used to store analysis results in ``trafo_data``.
+    """
+
+    _key = 'GlobalVarOffloadTransformation'
+
+    def __init__(self, key=None):
+        if key:
+            self._key = key
+
+    def transform_module(self, module, **kwargs):
+        """
+        Add device-side declarations for imported variables.
+        """
+
+        item = kwargs['item']
+
+        # bail if not global variable import
+        if not isinstance(item, GlobalVarImportItem):
+            return
+
+        # confirm that var to be offloaded is declared in module
+        symbol = item.local_name
+        assert symbol in [s.name.lower() for s in module.variables]
+
+        if not item.trafo_data.get(self._key, None):
+            item.trafo_data[self._key] = {'var_set': set()}
+
+        # do nothing if var is a parameter
+        if module.symbol_map[symbol].type.parameter:
+            return
+
+        # check if var is already declared
+        pragmas = [p for p in FindNodes(Pragma).visit(module.spec) if p.keyword.lower() == 'acc']
+        acc_pragma_parameters = get_pragma_parameters(pragmas, starts_with='declare', only_loki_pragmas=False)
+        if acc_pragma_parameters:
+            if symbol in flatten([v.replace(' ','').lower().split(',') for v in acc_pragma_parameters['create']]):
+                return
+
+        # Update the set of variables to be offloaded
+        item.trafo_data[self._key]['var_set'].add(symbol)
+
+        # Write ACC declare pragma
+        module.spec.append(Pragma(keyword='acc', content=f'declare create({symbol})'))
+
+    def transform_subroutine(self, routine, **kwargs):
+
+        role = kwargs.get('role')
+        item = kwargs['item']
+
+        # Bail if this routine is not part of a scheduler traversal
+        if item and not item.local_name == routine.name.lower():
+            return
+
+        if not item.trafo_data.get(self._key, None):
+            item.trafo_data[self._key] = {}
+
+        # Initialize sets/maps to store analysis
+        item.trafo_data[self._key]['modules'] = {}
+        item.trafo_data[self._key]['enter_data_copyin'] = set()
+        item.trafo_data[self._key]['enter_data_create'] = set()
+        item.trafo_data[self._key]['exit_data'] = set()
+        item.trafo_data[self._key]['acc_copyin'] = set()
+        item.trafo_data[self._key]['acc_copyout'] = set()
+
+        successors = kwargs.get('successors', ())
+        if role == 'driver':
+            self.process_driver(routine, successors)
+        if role == 'kernel':
+            self.process_kernel(routine, successors, item, kwargs['targets'])
+
+    def process_driver(self, routine, successors):
+        """
+        Add offload and/or copy-back directives for the imported variables.
+        """
+
+        update_device = ()
+        update_host = ()
+
+        # build offload pragmas
+        _acc_copyin = reduce(operator.or_, [s.trafo_data[self._key]['acc_copyin']
```

Comment:
`reduce` + `operator.or_` is actually unnecessary, you could just use `set.union` and tuple unpacking - e.g.:
```python
_acc_copyin = set.union(*[s.trafo_data.get(self._key, {}).get('acc_copyin', set()) for s in successors])
```

Reply:
Tips like these are highly appreciated; there are still many holes in my python knowledge. Thanks a lot 😄